fpga/lib/card/platform_card_factory.cpp

```diff
@@ -0,0 +1,131 @@
+
+using namespace villas;
+using namespace villas::kernel;
+using namespace villas::fpga;
+
+std::shared_ptr<PlatformCard>
+PlatformCardFactory::make(json_t *json_card, std::string card_name,
+                          std::shared_ptr<kernel::vfio::Container> vc,
+                          const std::filesystem::path &searchPath) {
+  auto logger = villas::logging.get("PlatformCardFactory");
+
+  json_t *json_ips = nullptr;
+  json_t *json_paths = nullptr;
+  const char *pci_slot = nullptr;
+  const char *pci_id = nullptr;
+  int do_reset = 0;
+  int affinity = 0;
+  int polling = 0;
+
+  json_error_t err;
+  int ret = json_unpack_ex(
+      json_card, &err, 0, "{ s: o, s?: i, s?: b, s?: s, s?: s, s?: b, s?: o }",
+      "ips", &json_ips, "affinity", &affinity, "do_reset", &do_reset, "slot",
+      &pci_slot, "id", &pci_id, "polling", &polling, "paths", &json_paths);
+
+  if (ret != 0)
+    throw ConfigError(json_card, err, "", "Failed to parse card");
+
+  CardParser parser(json_card);
+
+  auto card = std::make_shared<fpga::PlatformCard>(vc, parser.device_names);
+  card->name = std::string(card_name);
+  card->affinity = parser.affinity;
+  card->doReset = parser.do_reset != 0;
+  card->polling = (parser.polling != 0);
+
+  card->connectVFIOtoIPS();
+
+  // Load IPs from a separate json file
+  if (!json_is_string(json_ips)) {
+    logger->debug("FPGA IP cores config item is not a string.");
+    throw ConfigError(json_ips, "node-config-fpga-ips",
+                      "FPGA IP cores config item is not a string.");
+  }
+  if (!searchPath.empty()) {
+    std::filesystem::path json_ips_path =
+        searchPath / json_string_value(json_ips);
+    logger->debug("searching for FPGA IP cors config at {}",
+                  json_ips_path.string());
+    json_ips = json_load_file(json_ips_path.c_str(), 0, nullptr);
+  }
+  if (json_ips == nullptr) {
+    json_ips = json_load_file(json_string_value(json_ips), 0, nullptr);
+    logger->debug("searching for FPGA IP cors config at {}",
+                  json_string_value(json_ips));
+    if (json_ips == nullptr) {
+      throw ConfigError(json_ips, "node-config-fpga-ips",
+                        "Failed to find FPGA IP cores config");
+    }
+  }
+
+  if (not json_is_object(json_ips))
+    throw ConfigError(json_ips, "node-config-fpga-ips",
+                      "FPGA IP core list must be an object!");
+
+  ip::CoreFactory::make(card.get(), json_ips);
+  if (card->ips.empty())
+    throw ConfigError(json_ips, "node-config-fpga-ips",
+                      "Cannot initialize IPs of FPGA card {}", card_name);
+
+  // Additional static paths for AXI-Steram switch
+  if (json_paths != nullptr) {
+    if (not json_is_array(json_paths))
+      throw ConfigError(json_paths, err, "",
+                        "Switch path configuration must be an array");
+
+    size_t i;
+    json_t *json_path;
+    json_array_foreach(json_paths, i, json_path) {
+      const char *from, *to;
```

Comment:
indentation

Reply:
The file is formatted with clang-format.